### PR TITLE
Safe Mode implementation

### PR DESCRIPTION
**Pull request by Jamextreme140**
It disables the usage of classes that can perform malicious operations like Sys.command() and cpp.Lib.load().

**Reflect** and **Type** are disallowed since it can access to blacklisted classes (Reflect.callMethod() and Type.resolveClass())

Example: A song script tries to execute this following song script (function postCreate()):
`trace(Reflect.getProperty(PlayState, "difficulty"))`
`Sys.command("echo Howdy")`

Captura de pantalla 2024-06-01 124626
After disabled "Safe mode" it works normally again

_Safe Mode mark when it's enabled_
Captura de pantalla 2024-06-01 124528

### CODE REVIEW

**Comment by NeeEoo** [comment]
while(Script.importBlocklist.length > 0) Script.importBlocklist.pop()

**Comment by NexIsDumb** [comment]
and i'm also not sure about this either, well also first of all cause its easily avoidable as neeo made you see but also cause it would create MANY limitations

**Comment by Jamextreme140** [comment]
Wait. I think I have a better idea for Safe Mode

**Comment by NeeEoo** [comment]
We came up with a idea where like  its gonna be like a toggle thats by default off

**Comment by Jamextreme140** [comment]
Yeah, that's exactly what I was thinking about. Safe Mode as an Option instead of a separated build

**Comment by usb-port-2** [comment]
This _**SUCKS**_ so _**GODDAMN MUCH**_.

I should be able to _**DESTROY**_ your **_PC_** via my **_SILLY_** and **_FUNNY_** rhythm game song.

**Comment by Jamextreme140** [comment]
> We came up with a idea where like its gonna be like a toggle thats by default off

check you the latest commit

File: source/funkin/backend/scripting/Script.hx [hunk 1]
```diff
@@ -194,6 +205,8 @@ class Script extends FlxBasic implements IFlxDestroyable {
 		this.path = path;
 		onCreate(path);
 		for(k=>e in getDefaultVariables(this)) {
+			if(getBlacklistedImports().contains(k))
```
Comment (NeeEoo) [inline source/funkin/backend/scripting/Script.hx:208]: Id recommend calling this once just outside of the loop

File: source/funkin/backend/scripting/HScript.hx [hunk 2]
```diff
@@ -38,6 +38,10 @@ class HScript extends Script {
 		interp.importFailedCallback = importFailedCallback;
 		interp.staticVariables = Script.staticVariables;
 		interp.allowStaticVariables = interp.allowPublicVariables = true;
+		if(Options.safemode)
+		{
+			interp.importBlocklist = Script.getBlacklistedImports();
```
Comment (NeeEoo) [inline source/funkin/backend/scripting/HScript.hx:43]: Btw

`__script__.interp.importBlocklist` is accessable in a script

File: source/funkin/backend/scripting/Script.hx (hunk 1 above)
Comment (Jamextreme140) [reply source/funkin/backend/scripting/Script.hx:208]: Do you mean creating a separate array to use it or create a map from getDefaultVariables and modify it?

File: source/funkin/backend/scripting/HScript.hx (hunk 2 above)
Comment (Jamextreme140) [reply source/funkin/backend/scripting/HScript.hx:43]: well, there is the problem. I see it's not possible to deny the access to "importBlocklist" without breaking the entire engine.

File: source/funkin/backend/scripting/Script.hx (hunk 1 above)
Comment (NeeEoo) [reply source/funkin/backend/scripting/Script.hx:208]: Nah just like

var blockedImports = Script.getDefaultImports()
for k >= in aaaaaaa:
Im tooo tired rn but you get me

**Comment by NeeEoo** [comment]
We could probably just do @:unreflective on hscript-improved to turn off being able to edit it

**Comment by Jamextreme140** [comment]
> We could probably just do @:unreflective on hscript-improved to turn off being able to edit it

Hmm, it can work

**Pull request closed without merging**